syntax = "proto3";

service StorageServer {
    rpc Create(CreateRequest) returns (CreateResponse) {}
    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
    rpc Read(ReadRequest) returns (ReadResponse) {}
    rpc Write(WriteRequest) returns (WriteResponse) {}

    rpc GetFileList(GetFileListRequest) returns (GetFileListResponse) {}
}

message CreateRequest {
    string path = 1;
    string filename = 2;
    string content = 3;
}

message CreateResponse {
    bool success = 1;
}

message DeleteRequest {
    string path = 1;
    string filename = 2;
}

message DeleteResponse {
    bool success = 1;
}

message ReadRequest {
    string path = 1;
    string filename = 2;
}

message ReadResponse {
    string status = 1;
    string content = 2;
}

message WriteRequest {
    string path = 1;
    string filename = 2;
    string content = 3;
    int32 mode = 4; // 0: overwrite, 1: append
}

message WriteResponse {
    bool success = 1;
}


message GetFileListRequest {
    string path = 1;
}

message GetFileListResponse {
    message File {
        string filename = 1;
        string path = 2;
    }
    repeated File file_list = 1;
}
